onnxruntime/test/python/onnxruntime_test_collective.py

```diff
@@ -223,44 +232,50 @@ def test_all_gather_axis1(self):
 
         np.testing.assert_allclose(outputs[0], expected_output, err_msg="all gather on axis1: result mismatch")
 
-    def test_all_to_all(self):
-        for np_elem_type, elem_type, communication_elem_type in (
+    @parameterized.expand(
+        [
             (np.float32, TensorProto.FLOAT, TensorProto.FLOAT),
             (np.int64, TensorProto.INT64, TensorProto.INT64),
-            # TODO: Fix the following case, which triggers random number-mismatch error.
-            # (np.float32, TensorProto.INT64, TensorProto.BOOL),
-        ):
-            model = self._create_alltoall_ut_model(
-                (128, 128), elem_type=elem_type, communication_elem_type=communication_elem_type
-            )
-            rank, size = self._get_rank_size()
-            ort_sess = ort.InferenceSession(
-                model.SerializeToString(),
-                providers=["CUDAExecutionProvider", "CPUExecutionProvider"],
-                provider_options=[{"device_id": str(rank)}, {}],
-            )
+            (np.int64, TensorProto.INT64, TensorProto.BOOL),
+        ]
+    )
+    def test_all_to_all(self, np_elem_type, elem_type, communication_elem_type):
+        model = self._create_alltoall_ut_model(
+            (128, 128), elem_type=elem_type, communication_elem_type=communication_elem_type
+        )
+        rank, size = self._get_rank_size()
+        ort_sess = ort.InferenceSession(
+            model.SerializeToString(),
+            providers=["CUDAExecutionProvider", "CPUExecutionProvider"],
+            provider_options=[{"device_id": str(rank)}, {}],
+        )
 
+        if communication_elem_type == TensorProto.BOOL:
+            input = np.ones((128, 128), dtype=np_elem_type) if rank == 0 else np.zeros((128, 128), dtype=np_elem_type)
+
+            expected_output = np.concatenate(
+                (
+                    np.ones((128 // size, 128), dtype=np_elem_type),
+                    np.zeros((128 - (128 // size), 128), dtype=np_elem_type),
+                )
+            )
+        else:
             input = np.ones((128, 128), dtype=np_elem_type) * rank
-            outputs = ort_sess.run(None, {"X": input})
 
-            expected_output = np.zeros((int(128 / size), 128), dtype=np_elem_type)
+            expected_output = np.zeros((128 // size, 128), dtype=np_elem_type)
             for _ in range(size - 1):
                 expected_output = np.concatenate(
-                    (expected_output, np.ones((int(128 / size), 128), dtype=np_elem_type) * (_ + 1))
+                    (expected_output, np.ones((128 // size, 128), dtype=np_elem_type) * (_ + 1))
```

Comment:
Replace this with a formal python implementation of AllToAll. The  computation here is hard to understand. Maybe also refactorize the test following this type:
```python
tensor_on_device_0 = ...
tensor_on_deivce_1 = ...
...

if rank == 0:
  tensor_to_fed_to_ort = tensor_on_device_0
elif rank == 1:
  tensor_to_fed_to_ort = tensor_on_device_1
...

out = ort_session.run(tensor_to_fed_to_ort)

baseline_out = all_to_all_implementation(tensor_on_device_0, tensor_on_device_1, tensor_on_device_2, ...)

assert_all_close(baseline_out[rank], out)
```

Reply:
Your approach would create tensors for each device adding to the confusion. I have added explicit comments to state what the inputs are and the expected output and why.